src/wallet.cpp

```diff
@@ -4034,16 +4052,22 @@ void CWallet::AutoCombineDust()
         CScript scriptPubKey = GetScriptForDestination(it->first.Get());
         vecSend.push_back(make_pair(scriptPubKey, nTotalRewardsValue));
 
+        //Send change to same address
+        CTxDestination destMyAddress;
+        if (!ExtractDestination(scriptPubKey, destMyAddress)) {
+            LogPrintf("AutoCombineDust: failed to extract destination\n");
+            continue;
+        }
+        coinControl->destChange = destMyAddress;
+
         // Create the transaction and commit it to the network
         CWalletTx wtx;
         CReserveKey keyChange(this); // this change address does not end up being used, because change is returned with coin control switch
         string strErr;
         CAmount nFeeRet = 0;
 
-        //get the fee amount
-        CWalletTx wtxdummy;
-        CreateTransaction(vecSend, wtxdummy, keyChange, nFeeRet, strErr, coinControl, ALL_COINS, false, CAmount(0));
-        vecSend[0].second = nTotalRewardsValue - nFeeRet - 500;
+        // 10% safety margin to avoid "Insufficient funds" errors
+        vecSend[0].second = nTotalRewardsValue - (nTotalRewardsValue / 10);
```

Comment:
So this sets the fee to 10% of the transaction value? Not sure that this will work in all situations, such as when 10% of a very small value would be below the minimum tx fee. Maybe we should grab the serialized size of the tx and calculate fee per byte, then add a small safety net on top?

Reply:
No, it reduces the transaction value(!) by 10%, not the fee. And with the reduced 10% the insufficient funds error does not happen.

However, @Warrows removed the `CreateTransaction(vecSend, wtxdummy, keyChange, nFeeRet, strErr, coinControl, ALL_COINS, false, CAmount(0));` call which computes the fees with a dummy transaction before the actual transaction is created.
As it is now it's always created with `nFeeRet == 0`, I don't think that's correct.

I think a better way would be:
```
CAmount nFeeRet = 0;
//get the fee amount
CWalletTx wtxdummy;

// 10% safety margin to avoid "Insufficient funds" errors
vecSend[0].second = nTotalRewardsValue - (nTotalRewardsValue / 10);

// Compute fees
CreateTransaction(vecSend, wtxdummy, keyChange, nFeeRet, strErr, coinControl, ALL_COINS, false, CAmount(0));

// Subtract fees if possible
if(nTotalRewardsValue >= nFeeRet)
    vecSend[0].second = nTotalRewardsValue - nFeeRet;

if (!CreateTransaction(vecSend, wtx, keyChange, nFeeRet, strErr, coinControl, ALL_COINS, false, CAmount(0))) {
...
```

Reply:
Actually, the input value is unchanged, but the redeemed value is reduced by 10%.
It could be seen as 10% fees. But in fact one of the changes is to collect the value not needed for fees and put it back in the output. 

As pointed out by @Mrs-X  I also removed the evaluation of fees and send a transaction requiring 0 fee. This result in the `createTransaction()` function taking the minimal fee, up to 10% of the input. But as I said what is unused is returned. 

So the fees are actually left to the `createTransaction()` function to choose, as designed. It's now only called once.

One of the changes introduced in #497 consist in not sending transactions when the value is below the threshold and the fee not 0. So low value txs with fees above 10% should not be a concern.